### DISCUSSION TITLE
a demo based on VUE

### DISCUSSION THREAD

**Discussion by xu123shuai456** [yume-chan/ya-webadb#552] [Q&A] [answered]
Does this project have a demo based on VUE deployment?

**Comment by yume-chan** [comment]
I personally don't know Vue, but some forks have used this package with Vue, for example https://github.com/restsend/andbrowser

**Reply by xu123shuai456** [reply]
我需要调用那些API才能做到执行shell命令？
您可以帮忙贴一下最少的代码实现吗

**Reply by yume-chan** [reply]
I see, the demo has become too complex.

```js
import { Adb } from "@yume-chan/adb";
import { AdbWebUsbBackendManager } from "@yume-chan/adb-backend-webusb";
import AdbWebCredentialStore from "@yume-chan/adb-credential-web";

const CredentialStore = new AdbWebCredentialStore();

const button = document.getElementById("button");
button.addEventListener("click", async () => {
  if (!AdbWebUsbBackendManager.BROWSER) {
    alert("WebUSB is not supported in this browser");
    return;
  }

  const backend = await AdbWebUsbBackendManager.BROWSER.requestDevice();
  if (!backend) {
    // no device chosen by user
    return;
  }

  const connection = await backend.connect();

  // tell user to accept connection on device
  const device = await Adb.authenticate(
    connection,
    CredentialStore,
    undefined
  );
  const result = await device.subprocess.spawnAndWait("echo 'Hello, World!'");
  console.log(result.stdout);
});
```

**Reply by xu123shuai456** [reply]
非常感谢，它真的很好用。随便提个小建议，你有计划做一些类似实现最小功能的demo吗？
比如说实现执行adb shell的最小demo，实现adb push/pull的最小demo等

**Reply by yume-chan** [reply]
Thank you. There are also people complain about the demo being hard to use, so I have plan to make the demo more user-friendly like a real product, then create a new demo focusing on each feature and API usage using storybook UI.

**Reply by xu123shuai456** [reply]
hello，还有一个问题请教一下，这个用公网IP访问的，在其他电脑上用adb的时候，需要做什么特殊操作吗？

**Reply by xu123shuai456** [reply]
TypeError: Cannot read properties of undefined (reading 'requestDevice')
我在使用公网IP访问的时候，出现了这个错误

**Reply by yume-chan** [reply]
https://github.com/yume-chan/ya-webadb/blob/main/CONTRIBUTING.md#1-webusb-and-file-downloading-doesnt-work-in-developmentself-host-environment

**Reply by xu123shuai456** [reply]
这个是生成静态文件的？好像不是我遇到的错误。
console.log(window.navigator.usb)
这个通过http://localhost:5173/访问的时候，可以打印出来值，通过公网IP访问没有
我不太清楚是否与AdbProxyServer有关，但是我无法安装这个库

**Reply by yume-chan** [reply]
Read the bottommost section.

**Reply by xu123shuai456** [reply]
非常感谢，是我理解错了。
我添加了自定义的ssl证书，好使了。棒

**Reply by xu123shuai456** [reply]
hello，请问这个库有办法获得手机root权限吗？

**Reply by yume-chan** [reply]
If your device is already rooted, you can start adbd as root or run all commands with `su` in Tango.

If you want to root your device, I'm not aware of any methods that works on recent Android versions.

**Reply by Muthu-Palaniyappan-OL** [reply]
This is works only for 0.19 its not available in 0.20

**Reply by yume-chan** [reply, marked as answer]
For 0.0.20:

```ts
import { Adb, AdbDaemonTransport } from "@yume-chan/adb";
import AdbWebCredentialStore from "@yume-chan/adb-credential-web";
import { AdbDaemonWebUsbDeviceManager } from "@yume-chan/adb-daemon-webusb";

const CredentialStore = new AdbWebCredentialStore();

const button = document.getElementById("button");
button.addEventListener("click", async () => {
  if (!AdbDaemonWebUsbDeviceManager.BROWSER) {
    alert("WebUSB is not supported in this browser");
    return;
  }

  const backend = await AdbDaemonWebUsbDeviceManager.BROWSER.requestDevice();
  if (!backend) {
    // no device chosen by user
    return;
  }

  const connection = await backend.connect();

  // tell user to accept connection on device
  const device = new Adb(
    await AdbDaemonTransport.authenticate({
      serial: backend.serial,
      connection,
      credentialStore: CredentialStore,
    })
  );
  const result = await device.subprocess.spawnAndWait("echo 'Hello, World!'");
  console.log(result.stdout);
});
```

**Reply by meetAndEgg** [reply]
Can you add scrcpy to this demo? The project is a little hard for me.
